Add algorithm selection for training/importing

Corentin-pro opened the pull request:
I haven't check the test yet but I think this is a good base to see the direction of this feature and discuss. Everything is working from my manual tests but I might missed some things.

Resolves #327 

Summary of the commit :
* Add 'algorithm' key in aiengine protos (StartTrainingRequest, ImportModelRequest)
* Add 'algorithm' key in runtime protos (TrainModel, ImportModel)
* Clean code (PEP8, flake) for ai scripts : main.py and train.py
* Reduce Tensorflow verbose before import to avoid unecessary info logs
  (main.py l.318)
* Add 'algorithm' flag for import and train command (CLI)

Commit:
Add algorithm selection for training/importing

* Add 'algorithm' key in aiengine protos (StartTrainingRequest, ImportModelRequest)
* Add 'algorithm' key in runtime protos (TrainModel, ImportModel)
* Clean code (PEP8, flake) for ai scripts : main.py and train.py
* Reduce Tensorflow verbose before import to avoid unecessary info logs
  (main.py l.318)
* Add 'algorithm' flag for import and train command (CLI)

lukekim commented on ai/src/algorithms/factory.py:
```diff
@@ -6,7 +6,8 @@
+    if name == "dql":
         return DeepQLearning_Agent(state_shape, action_size)
 
     raise NotImplementedError(f"Unable to find agent with name '{name}'")
```
Maybe we should update this error message to refer to learning-algorithm?

Corentin-pro replied:
The message has been update.

lukekim commented on pkg/cli/cmd/train.go:
```diff
@@ -82,7 +85,16 @@ spice train logpruner.yaml
-		response, err := http.DefaultClient.Post(trainUrl, "application/json", nil)
+
+		trainRequest := &runtime_pb.TrainModel{
+			Algorithm: algorithmFlag,
```
As we want to be developer friendly as possible, let's call this `LearningAlgorithm` and in json `learning_algorithm` because many developers will not understand what this algorithm refers to.

Corentin-pro replied:
Agreed, I changed all occurrences in APIs/CLI flags to `learning_algorithm`. Only short lived local variables remains with the name `algorithm` and the meta data of the model.

lukekim commented on pkg/cli/cmd/train.go:
```diff
@@ -117,5 +129,6 @@ spice train logpruner.yaml
 
 func init() {
 	trainCmd.Flags().StringVar(&contextFlag, "context", "docker", "Runs Spice.ai in the given context, either 'docker' or 'metal'")
+	trainCmd.Flags().StringVar(&algorithmFlag, "algorithm", "", "Train the pod with specified algorithm")
```
same comment as above, many users won't know what algorithm refers to, so let's use `learning-algorithm`

lukekim commented on pkg/cli/cmd/import.go:
```diff
@@ -76,5 +77,6 @@ spice import --tag latest ./models/trader.spicepod
 
 func init() {
 	ImportCmd.Flags().StringVar(&importTag, "tag", "latest", "Specify which tag to import the model to")
+	ImportCmd.Flags().StringVar(&algorithmTag, "algorithm", "dql", "Specify algorithm of the pod")
```
```suggestion
	ImportCmd.Flags().StringVar(&algorithmTag, "learning-algorithm", "dql", "Specify the AI engine learning algorithm")
```

lukekim commented:
Nice work! You'll also need to update the `.github/worflows/e2e.yaml` with the algorithm change.

lukekim commented on pkg/pods/podparams.go:
```diff
@@ -7,12 +7,14 @@ type PodParams struct {
 		Period:      time.Hour * 24 * 3,
 		Interval:    time.Minute * 1,
 		Granularity: time.Second * 10,
+		Algorithm: "dql",
```
Looks like you might need to run `gofmt -w -s -l .`

Corentin-pro replied:
Indeed, I pushed a commit fixing this, next time I will set up my IDE to format on save.

Commit:
Saving model with algorithm information

* Fix code format (linter OK)

Commit:
All tests passing

* Fix tensorflow eager mode not bein disabled during tests
* Fix agent save/load

Commit:
Rename Algorithm to LearningAlgorithm in the API

lukekim commented on ai/src/Makefile:
```diff
@@ -1,7 +1,7 @@
 test: venv
-	SPICE_DEEPRL_ALGORITHM="vpg" venv/bin/python3 -m pytest --timeout 10
-	SPICE_DEEPRL_ALGORITHM="dql" venv/bin/python3 -m pytest --timeout 10
+	ALGORITHM="dql" venv/bin/python3 -m pytest --timeout 10
```
One reason to prefix with `SPICE_` is the environment variable can be used in the manifest, as the config will auto-replace anything with that prefix.

Corentin-pro replied:
OK, I didn't know that, the good thing is that this environment variable is exclusively used for pytest as I couldn't figured out yet how to parameterize the testing (without a lot of changes).

A solution here would be to prefix with 'TEST_' to explicit the nature of this environment variable.

Commit:
Fix pod Import

* Change meta data file's model_path (absolute) to model_name (file name
  only) when saving model as path after import wil be different
* Remove LearningAlgorithm from ImportRequest and --learning-algorithm
  from CLI import as the information is read from meta data

Corentin-pro commented:
@phillipleblanc I made the change to pass all the e2e test, I do not understand the pr_test well (`go test -v ./... -shuffle=on`). I feel I will need some guidance to fix it.

phillipleblanc commented:
> @phillipleblanc I made the change to pass all the e2e test, I do not understand the pr_test well (`go test -v ./... -shuffle=on`). I feel I will need some guidance to fix it.

Looks like this change will fix the Go unit tests.

<img width="764" alt="Screen Shot 2021-10-20 at 6 51 57 PM" src="https://user-images.githubusercontent.com/879445/138072575-8236d60a-21bd-431a-a800-540bb5ecc64f.png">

Commit:
Fix e2e test to use new --learning-algorithm flag

* Fix snapshot test (thanks to Philip)

Corentin-pro committed:
Merge branch 'trunk' into learning_algorithm_param

Corentin-pro commented:
Now every test is OK!

The new feature works as follow :

- '--learning-algorithm' can be used with the `train` command of the CLI (like `spice train [pod] --learning-algorithm [dql|vpg]`)
- a pod can specify a default learning_algorithm in its yaml (I need to check this more, I don't think there is a test for that)
- to avoid problem with import/export (I made the same flag there at first but eventually this was not a good solution) the model is now saved as a folder with a `meta.json` with meta data and `model.pb`. For now the meta data only consist of `algorithm` and 'model_name' (storing `model.pb`, could change latter).

As a side note I changed the temporary folder creation by adding a prefix (the python function that create the temporary folder had the parameter but was unused) to start with `spiceai_`. In case of crashes/wrong cleanup this is easier to clean manually and cleaner overall.

phillipleblanc reviewed (dismissed):
Awesome first PR! Thanks for working through those test issues.

Commit:
Format go source files (go fmt ./...)
